### PR TITLE
Random and word based string endpoint

### LINKED ISSUE #13: Random String endpoint
- max
- min
- format (ex. Uppercase. PascalCase)
- allowed characters
- encoding (Base64)

### PR DESCRIPTION
**Pull request by KanarekLife**

### CODE REVIEW

File: RandomData.Api.Tests/RandomData.Api.Tests.csproj [hunk 1]
```diff
@@ -10,6 +10,7 @@
     <PackageReference Include="Moq" Version="4.14.1" />
     <PackageReference Include="xunit" Version="2.4.1" />
     <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.0.0" />
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/RandomData.Api.Tests.csproj]: Please, use the newest version: 16.6.1

File: RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿using FluentAssertions;
+using Microsoft.Extensions.DependencyInjection;
+using RandomData.Api.Services.FileReader;
+using RandomData.Api.Services.FileReader.ServiceImplementations;
+using Xunit;
+
+namespace RandomData.Api.Tests.Services.FileReaderService
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs]: In directory and namespaces in test project are FileReaderService but in production code are FileReader. Probably you forgot to rename after other changes.

File: RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+{
+    public class FileReaderServiceHelpersTests
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs]: In TDD or 100% practices is cool test, but in most cases it is not necessary does not add too much value.

File: RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        [Fact]
+        public void IsAddFileReaderWorkingProperly()
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/Services/FileReaderService/FileReaderServiceHelpersTests.cs]: maybe instead of "IsAddFileReaderWorkingProperly" you can use "IsAddFileReaderRegisteredCorrectly"

File: RandomData.Api.Tests/Services/RandomService/RandomServiceHelpersTest.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        [Fact]
+        public void IsAddRandomWorkingProperly()
+        {
+            //arrange
+            var serviceProvider = new ServiceCollection()
+                .AddRandomService()
+                .BuildServiceProvider();
+
+            //act/assert
+            serviceProvider.GetService<IRandom>().Should().BeOfType<Random>();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/Services/RandomService/RandomServiceHelpersTest.cs]: Your act and assert is just act, but arrange is arrange and act. Another your test is (IsAddFileReaderWorkingProperly) correct.

Try to add to act only tested action. To assert add all actions that allows you to validate results, it's ok to use another method that gives you data. Everything else should be in arrange.

File: RandomData.Api.Tests/StringGenerationTests/Extensions/StringEncodingExtensionMethodsTests.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+using Xunit;
+
+namespace RandomData.Api.Tests.StringGenerationTests.Extensions
+{
+    public class StringEncodingExtensionMethodsTests
+    {
+        [Theory]
+        [InlineData(Encoding.Base64, "aAaA", "YUFhQQ==")]
+        [InlineData(Encoding.Default, "aAaA", "aAaA")]
+        public void EncodeToExtensionTests(Encoding encoding, string input, string output)
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/Extensions/StringEncodingExtensionMethodsTests.cs]: All public method in this class are tests. You don't have to add "tests" postfix.

File: RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/RandomStringGenerationServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        [InlineData(Format.Snake, "hello_world")]
+        [InlineData(Format.Upper, "HELLO WORLD")]
+        public void FormatPropertyTest(Format format, string expectedOutput)
+        {
+            //arrange
+            var random = new FakeRandom(new[] {11, 0, 1, 2, 2, 3, 4, 5, 3, 6, 2, 7});
+            var service = new RandomStringGenerationService(random);
+
+            //act
+            var result = service.GenerateRandomString(new GetStringParameters
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/RandomStringGenerationServiceTests.cs]: To simplify act section is nice to create variable in arrange and just use it here. It's not necessary just an opinion.

File: RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/RandomStringGenerationServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+        [Fact]
+        public void ShouldReturnInvalidParametersException()
+        {
+            //arrange
+            var random = new Random();
+            var service = new RandomStringGenerationService(random);
+
+            //act/assert
+            service.Invoking(x => x.GenerateRandomString(new GetStringParameters
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/RandomStringGenerationServiceTests.cs]: You can create Action variable in act version and use it in assert. It's not necessary just an opinion.

File: RandomData.Api/appsettings.json [hunk 9]
```diff
@@ -7,6 +7,9 @@
     }
   },
   "AllowedHosts": "*",
+  "StringGenerationOptions" : {
+    "WordsDictionaryLocation" : "words.json"
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/appsettings.json]: Please, add sample words.json file to project and mark to copy always to output directory.
Or even better add file with real Polish words :)

File: RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/WordsStringGenerationServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        [Theory]
+        [InlineData((string) null)]
+        [InlineData("")]
+        [InlineData("{}")]
+        [InlineData("[Hamburger]")]
+        public void ConstructorShouldReturnInvalidWordsDictionaryExceptionOnInvalidJson(string content)
+        {
+            //arrange
+            var fakeFile = new FakeFileReader(content);
+            Action a = () => new WordsStringGenerationService(GetFakeOptions(), fakeFile, new Random());
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/ServiceImplementations/WordsStringGenerationServiceTests.cs]: Please, do not use "a" as name of variable. Maybe "Construct/CreateWordsStringGenerationService".

File: RandomData.Api/Services/FileReader/FileReaderHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using RandomData.Api.Services.FileReader.ServiceImplementations;
+
+namespace RandomData.Api.Services.FileReader
+{
+    public static class FileReaderHelpers
+    {
+        public static IServiceCollection AddFileReaderService(this IServiceCollection serviceCollection)
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Services/FileReader/FileReaderHelpers.cs]: Common name for this parameter is services.

File: RandomData.Api/Services/FileReader/ServiceImplementations/SystemFileReader.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
+﻿using System.IO;
+
+namespace RandomData.Api.Services.FileReader.ServiceImplementations
+{
+    public class SystemFileReader : IFileReader
+    {
+        public string GetFileContent(string path)
+        {
+            if (!File.Exists(path) || string.IsNullOrWhiteSpace(path)) throw new FileNotFoundException();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Services/FileReader/ServiceImplementations/SystemFileReader.cs]: I prefer a version: 
if(condition)
    expression;
It's more readable. You can easily see different code paths.

File: RandomData.Api/Services/Random/ServiceImplementations/Random.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+
+namespace RandomData.Api.Services.Random.ServiceImplementations
+{
+    public class Random : IRandom
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Services/Random/ServiceImplementations/Random.cs]: Maybe to avoid name collision we should change name to RandomGenerator

File: RandomData.Api/Startup.cs [hunk 14]
```diff
@@ -20,6 +23,9 @@ public Startup(IConfiguration configuration)
 		public void ConfigureServices(IServiceCollection services)
 		{
 			services.AddSwaggerWithConfig(Configuration.GetSection("SwaggerConfig").Get<OpenApiInfo>());
+      services.AddFileReaderService();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Startup.cs]: Code formatting

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+        public RandomStringController(RandomStringGenerationService stringGenerationService,
+            GetStringParametersValidator validator)
+        {
+            _stringGenerationService = stringGenerationService;
+            _validator = validator;
+        }
+
+        /// <summary>
+        ///     Returns random string
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: Unnecessary spaces in xml comment

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+namespace RandomData.Api.StringGeneration.Controllers
+{
+    [ApiController]
+    [Route("/string")]
+    public class RandomStringController : Controller
+    {
+        private readonly RandomStringGenerationService _stringGenerationService;
+        private readonly GetStringParametersValidator _validator;
+
+        public RandomStringController(RandomStringGenerationService stringGenerationService,
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: You should use interface

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            _stringGenerationService = stringGenerationService;
+            _validator = validator;
+        }
+
+        /// <summary>
+        ///     Returns random string
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet("random")]
+        [HttpGet("")]
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: I think empty routing is enough.

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+            GetStringParametersValidator validator)
+        {
+            _stringGenerationService = stringGenerationService;
+            _validator = validator;
+        }
+
+        /// <summary>
+        ///     Returns random string
+        /// </summary>
+        /// <returns></returns>
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: You can remove not used section like "returns" in this case.

File: RandomData.Api/StringGeneration/Dto/GetStringParameters.cs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+            " !\"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
+
+        /// <summary>
+        ///     Minimum length of created string (ignored if length is set); Default value is 1
+        /// </summary>
+        [FromQuery(Name = "minLength")]
+        public int MinLength { get; set; } = 1;
+
+        /// <summary>
+        ///     Maximum length of created string (ignored if length is set); Default value is 100
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Dto/GetStringParameters.cs]: "(ignored if length is set)" there is no length property anymore.

File: RandomData.Api/StringGeneration/Exceptions/InvalidParametersException.cs [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+using System.Linq;
+using FluentValidation.Results;
+
+namespace RandomData.Api.StringGeneration.Exceptions
+{
+    public class InvalidParametersException : Exception
+    {
+        public InvalidParametersException(IList<ValidationFailure> validationResultErrors) :
+            base("Parameters couldn't be validated. " +
+                 $"{validationResultErrors.Select(x => x.ToString()).Aggregate((s, s1) => s + "\n" + s1)}")
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/Exceptions/InvalidParametersException.cs]: You can do it simpler, by using string.Join
$"Parameters couldn't be validated. {string.Join(Environment.NewLine, validationResultErrors)}"

File: RandomData.Api/StringGeneration/StringGenerationServiceHelpers.cs [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            return services
+                .AddTransient<WordsStringGenerationService>()
+                .AddTransient<RandomStringGenerationService>()
+                .AddTransient<GetStringParametersValidator>()
+                .AddTransient(serviceProvider => serviceProvider.GetService<IConfiguration>()
+                    .GetSection("StringGenerationOptions").Get<StringGenerationServiceOptions>());
+        }
+
+        public class StringGenerationServiceOptions
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/StringGenerationServiceHelpers.cs]: Please, move to separate file.

File: RandomData.Api/StringGeneration/StringGenerationServiceHelpers.cs [hunk 22]
```diff
@@ -0,0 +1,25 @@
+﻿using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using RandomData.Api.StringGeneration.ServiceImplementations;
+using RandomData.Api.StringGeneration.Validators;
+
+namespace RandomData.Api.StringGeneration
+{
+    public static class StringGenerationServiceHelpers
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/StringGenerationServiceHelpers.cs]: By conventions name IServiceCollection extensions are created by adding postfix Extenssions to name of "module". StringGenerationExtensions is ok.

File: RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+using RandomData.Api.StringGeneration.Exceptions;
+using RandomData.Api.StringGeneration.Extensions;
+using RandomData.Api.StringGeneration.Validators;
+
+namespace RandomData.Api.StringGeneration.ServiceImplementations
+{
+    public class WordsStringGenerationService : IStringGenerationService
+    {
+        private readonly IRandom _random;
+        private readonly GetStringParametersValidator _validator = new GetStringParametersValidator();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs]: Use interface and inject in constructor to improve testability.

File: RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+        private readonly IEnumerable<string> _words;
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs]: You can use Array or even ISet/HashSet. More specific type gives extra information about usage ad purpose.

File: RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+        public WordsStringGenerationService(StringGenerationServiceHelpers.StringGenerationServiceOptions options,
+            IFileReader fileReader, IRandom random)
+        {
+            _random = random;
+            var path = options.WordsDictionaryLocation;
+            if (string.IsNullOrEmpty(path)) throw new WordsDictionaryLocationUnspecifiedException();
+            var content = fileReader.GetFileContent(path);
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs]: This service is injected as transient so in each request try to read content of file to memory. This behavior really slow down this endpoint.

File: RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+        public string GenerateRandomString(GetStringParameters parameters)
+        {
+            var validationResult = _validator.Validate(parameters);
+            if (!validationResult.IsValid) throw new InvalidParametersException(validationResult.Errors);
+
+            var length = _random.Next(parameters.MinLength, parameters.MaxLength);
+
+            var stringChars = new char[length];
+
+            for (var i = 0; i < length; i++)
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs]: Or new string(Enumerable.Range(0, length).Select(GetRandomChar()).ToArray());

File: RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        private readonly IRandom _random;
+        private readonly GetStringParametersValidator _validator = new GetStringParametersValidator();
+
+        public RandomStringGenerationService(IRandom random)
+        {
+            _random = random;
+        }
+
+        public string GenerateRandomString(GetStringParameters parameters)
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs]: If you group line of codes in groups by adding empty lines it's probably information that you should split this method to smaller.

File: RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+            serviceProvider.GetService<StringGenerationServiceHelpers.StringGenerationServiceOptions>()
+                .Should().BeEquivalentTo(expectedOptions);
+        }
+    }
+
+    internal static class HelpingExtensions
+    {
+        internal static Stream ToStream(this string s)
+        {
+            var stream = new MemoryStream();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs]: maybe just:
return new MemoryStream(Encoding.UTF8.GetBytes(s));

File: RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+        public void IsAddStringGenerationServicesWorkingProperly()
+        {
+            //arrange
+            var serviceCollection = new ServiceCollection();
+            var expectedOptions = new StringGenerationServiceHelpers.StringGenerationServiceOptions
+            {
+                WordsDictionaryLocation = "words.json"
+            };
+            IConfiguration configuration = new ConfigurationBuilder()
+                .AddJsonStream(
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs]: I think that content of file doesn't matter during initialization. You can add empty stream.

File: RandomData.Api.Tests/StringGenerationTests/Extensions/StringFormattingExtensionMethodsTests.cs [hunk 30]
```diff
@@ -0,0 +1,75 @@
+﻿using FluentAssertions;
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGenerationTests/Extensions/StringFormattingExtensionMethodsTests.cs]: Maybe good idea is adding also test with 2 spaces between characters: "aa  AA".

**Review by krzysztof-gwozdz — CHANGES_REQUESTED** [review]

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs (hunk 16 above)
Comment (KanarekLife) [reply RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: But when I use an interface then how can I specify between one of two implementations?

File: RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs (hunk 23 above)
Comment (KanarekLife) [reply RandomData.Api/StringGeneration/ServiceImplementations/WordsStringGenerationService.cs]: How can I use interface instead of GetStringParametersValidator?

File: RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs (hunk 29 above)
Comment (KanarekLife) [reply RandomData.Api.Tests/StringGenerationTests/StringGenerationServiceHelpersTests.cs]: Currently it does matter because the services depend on correct json config. Without it the exception is thrown.

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs (hunk 15 above)
Comment (KanarekLife) [reply RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: I personally think it improved readibility but if you insist I can of course change it.

File: RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs (hunk 27 above)
Comment (KanarekLife) [reply RandomData.Api/StringGeneration/ServiceImplementations/RandomStringGenerationService.cs]: I think that the length of this method is ok and there is no need to split it more. I would even say that it wouldn't improve a lot. However if you really think it should be changed I'll of course do it.

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs (hunk 15 above)
Comment (krzysztof-gwozdz) [reply RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: I don't know how more than one space can improve readability :P If somebody decided to add new xml comment and want to be constant with your solution, after auto-genereting comment must always add 4 spaces or tab. Sounds annoying for me. Maybe we should ask whole team.

File: RandomData.Api/WordGeneration/WordController.cs [hunk 31]
```diff
@@ -0,0 +1,35 @@
...
+        public WordController(GetWordParametersValidator validator, WordGeneration wordGeneration)
+        {
+            _validator = validator;
+            _wordGeneration = wordGeneration;
+        }
+
+        /// <summary>
+        ///     Returns random word
+        /// </summary>
+        /// <response code="200">Ok</response>
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/WordGeneration/WordController.cs]: Basically, you just translate status code to their names. Maybe 200 - Random word, 400 - Error message?

File: RandomData.Api/WordGeneration/WordController.cs [hunk 32]
```diff
@@ -0,0 +1,35 @@
...
+        /// <response code="401">Bad Request</response>
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/WordGeneration/WordController.cs]: 401 is not Bad Request

File: RandomData.Api/StringGeneration/StringController.cs [hunk 33]
```diff
@@ -0,0 +1,36 @@
+﻿using Microsoft.AspNetCore.Mvc;
+using RandomData.Api.StringGeneration.Dto;
+using RandomData.Api.StringGeneration.Validators;
+
+namespace RandomData.Api.StringGeneration
+{
+    [ApiController]
+    [Route("/string")]
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/StringGeneration/StringController.cs]: You can use [controller] instead of /string

File: RandomData.Api/Services/Random/RandomHelpers.cs [hunk 34]
```diff
@@ -0,0 +1,12 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+
+namespace RandomData.Api.Services.Random
+{
+    public static class RandomHelpers
+    {
+        public static IServiceCollection AddRandomService(this IServiceCollection serviceCollection)
+        {
+            return serviceCollection.AddTransient<IRandomGenerator, ServiceImplementations.RandomGenerator>();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Services/Random/RandomHelpers.cs]: Please move namespace to using section

File: RandomData.Api/Services/FileReader/ServiceImplementations/SystemFileReader.cs [hunk 35]
```diff
@@ -0,0 +1,17 @@
+﻿using System.IO;
+
+namespace RandomData.Api.Services.FileReader.ServiceImplementations
+{
+    public class SystemFileReader : IFileReader
+    {
+        public string GetFileContent(string path)
+        {
+            if (!File.Exists(path) || string.IsNullOrWhiteSpace(path))
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Services/FileReader/ServiceImplementations/SystemFileReader.cs]: Switch order. In this version if path is empty or white space that also means that it doesn't exist. Second condition is checked only if path exists so is not empty.

File: RandomData.Api/RandomData.Api.csproj [hunk 36]
```diff
@@ -12,7 +12,19 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.4.1" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.5.0" />
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/RandomData.Api.csproj:15]: Please do not update things that are not related or necessary to complete your task. In this case it can stay but with more significant changes something like this creates conflicts.

File: RandomData.Api/Extensions/StringManipulation/StringFormattingExtensions.cs [hunk 37]
```diff
@@ -0,0 +1,90 @@
...
+                        makeUpperCase = false;
+                        return char.ToUpper(x);
+                    }
+
+                    return char.ToLower(x);
+                })
+                .Where(x => x != ' ').ToArray());
+        }
+
+        public static string ToSnakeCase(this string input)
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/Extensions/StringManipulation/StringFormattingExtensions.cs:74]: What do you think about input?.Trim().ToLower().Replace(' ', '_');

File: RandomData.Api/WordGeneration/WordGeneration.cs [hunk 38]
```diff
@@ -0,0 +1,63 @@
...
+        private readonly ISet<string> _words;
+
+        public WordGeneration(WordsGenerationOptions options,
+            IFileReader fileReader, IRandomGenerator randomGenerator, GetWordParametersValidator validator, IMemoryCache memoryCache)
+        {
+            _randomGenerator = randomGenerator;
+            _validator = validator;
+            var path = options.WordsDictionaryLocation;
+            if (string.IsNullOrEmpty(path)) throw new WordsDictionaryLocationUnspecifiedException();
+            var result = memoryCache.TryGetValue(WordsDictionaryMemoryCacheKey, out _words);
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/WordGeneration/WordGeneration.cs]: It's constructor so probably is not method result. So what kind of result is it?

Comment (krzysztof-gwozdz) [reply RandomData.Api/WordGeneration/WordGeneration.cs]: Maybe this part should be moved to separate method to simplify constructor code.

File: RandomData.Api/WordGeneration/WordGeneration.cs [hunk 39]
```diff
@@ -0,0 +1,63 @@
...
+        public string GenerateRandomString(GetWordParameters parameters)
+        {
+            var validationResult = _validator.Validate(parameters);
+            if (!validationResult.IsValid) throw new InvalidParametersException(validationResult.Errors);
+            
+            var regex = new Regex($"[^{parameters.AllowedCharacters}]");
+            
+            var filteredWords = _words
+                .Where(x => x.Length >= parameters.MinLength && x.Length <= parameters.MaxLength)
+                .Where(x => !regex.IsMatch(x)).ToArray();
```
Comment (krzysztof-gwozdz) [inline RandomData.Api/WordGeneration/WordGeneration.cs]: ToArray to new line

File: RandomData.Api/StringGeneration/Controllers/RandomStringController.cs (hunk 15 above)
Comment (KanarekLife) [reply RandomData.Api/StringGeneration/Controllers/RandomStringController.cs]: I think that a simple solution for that will be to use some sort of linter as a part of CI and before that create a rules with [editor onfig](https://docs.microsoft.com/en-us/visualstudio/ide/editorconfig-code-style-settings-reference?view=vs-2019) . I used this convention due to rider and its default settings and I don't think it looks so terrible 😆

File: RandomData.Api.Tests/StringGeneration/StringGenerationTests.cs [hunk 40]
```diff
@@ -0,0 +1,136 @@
...
+
+        [Fact]
+        public void ShouldReturnInvalidParametersException()
+        {
+            //arrange
+            var random = new RandomGenerator();
+            var service = new Api.StringGeneration.StringGeneration(random, new GetStringParametersValidator());
+
+            //act/assert
+            service.Invoking(x => x.GenerateRandomString(new GetStringParameters
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/StringGeneration/StringGenerationTests.cs]: Please separate this to act and assert with action variable.

File: RandomData.Api.Tests/WordGeneration/WordsGenerationTests.cs [hunk 41]
```diff
@@ -0,0 +1,202 @@
...
+        [Theory]
+        [InlineData((string) null)]
+        [InlineData("")]
+        [InlineData("{}")]
+        [InlineData("[Hamburger]")]
+        public void ConstructorShouldReturnInvalidWordsDictionaryExceptionOnInvalidJson(string content)
+        {
+            //arrange
+            var fakeFile = new FakeFileReader(content);
+            Action a = () => new Api.WordGeneration.WordGeneration(GetFakeOptions(), fakeFile, new RandomGenerator(), new GetWordParametersValidator(), new MemoryCache(new MemoryCacheOptions()));
```
Comment (krzysztof-gwozdz) [inline RandomData.Api.Tests/WordGeneration/WordsGenerationTests.cs]: It's act part to arrange. a is not best name for variable Maybe CreateWordGeneration

**Review by krzysztof-gwozdz — APPROVED** [review]

**Review by mejtika — APPROVED** [review]
